File: RLCore.fs
module RLCore

open System

type StateIdx<'s> = 's->int
type ActionIdx<'a> = 'a->int
type BaseTransitionFunction<'s, 'a> = 's->'a->(('s * float) list)
type TransitionFunction<'s, 'a>= 's->'a->'s->float

let rand = Random()

let rec getRandomElementWeightedSub (lst: 'a list) (weigths: float list) (r: float): 'a =
    if (r - (List.head weigths)) > 0 then
        getRandomElementWeightedSub (List.tail lst) (List.tail weigths) (r - List.head weigths)
    else
        List.head lst

let getRandomElementWeighted (lst: 'a list) (weigths: float list): 'a =
    getRandomElementWeightedSub lst weigths (rand.NextDouble() * (List.sum weigths))

/// Konverterer en BaseTransitionFunciton til en almindelig transitionfunction til brug i value iteration
let convertTransitionFunction<'s, 'a when 's : equality> (bt: BaseTransitionFunction<'s, 'a>) : TransitionFunction<'s, 'a> =
    let tf (s1: 's) (a: 'a) (s2: 's) : float =
        let probs: ('s * float) list = bt s1 a
        let filtered = List.filter (fun x -> fst x = s2) probs
        match filtered with
        | n :: rest -> snd n
        | [] -> 0
    tf

// Dealing with Q-values

/// Given a state and the calculated Q-table get the value of the state
/// Which is the action giving the maximum value
let getStateValue<'s> (si: StateIdx<'s>) (s: 's) (Q: float[,]) : float =
    Array.max Q[si s, *]

/// Given a starting state an action and the end state what is the value
/// This is equal to the transition reward plus the discounted (expected) value of the end state
let getActionToStateValue<'s, 'a> (si: StateIdx<'s>) (ai: ActionIdx<'a>) (R: float[,,]) (discount: float) (stateFrom: 's) (action: 'a) (stateTo: 's) (Q: float[,]) : float =
    let transitionReward = R[si stateFrom, ai action, si stateTo]
    let nextStateValue = getStateValue si stateTo Q
    transitionReward + discount * nextStateValue

File: Python/RLCore.py
from typing import Any, Callable, TypeVar
import random as rnd

# A base transition function is a transition function that takes a state
# and an action and returns a list of tuples containing the ending states and
# the probability of ending in each of those (should sum to 1).
# Any state not in the list has probability 0 of being the destination

S = TypeVar('S')
A = TypeVar('A')

def getRandomElementWeightedSub (lst: 'list[S]', weigths: 'list[float]', r: float) -> S:
    if (r - weigths[0] > 0):
        return getRandomElementWeightedSub(lst[1:], weigths[1:], r - weigths[0])
    else:
        return lst[0]

def getRandomElementWeighted (lst: 'list[S]', weigths: 'list[float]') -> S :
    return getRandomElementWeightedSub(lst, weigths, rnd.random() * sum(weigths))

def baseTransitionFunctionToNormalTransitionFunction(base: 'Callable[[S, A], list[tuple[S, float]]]') -> 'Callable[[S, A, S], float]':
    def normal(sFrom: S, a: A, sTo: S) -> float:
        pairs = base(sFrom, a) # list of tuples with state and probability
        return sum([x[1] for x in pairs if x[0]== sTo])
    return normal

def baseTransitionFunctionToStochasticTransitionFunction(base: 'Callable[[S, A], list[tuple[S, float]]]') -> 'Callable[[S, A], S]':
    def tf(s1: S, a: A) -> S:
        probs = base(s1, a)
        return getRandomElementWeighted([x[0] for x in probs], [x[1] for x in probs])
    return tf

# Given a state and the calculated Q-table get the value of the state
# Which is the action giving the maximum value
def getStateValue(stateIdx: 'Callable[[S], int]', state: S, Q: 'list[list[float]]' ) -> float:
    return max(Q[stateIdx(state)])

# Given a starting state an action and the end state what is the value
# This is equal to the transition reward plus the discounted (expected) value of the end state
def getActionToStateValue(stateIdx: ('Callable[[S], int]'), actionIdx: 'Callable[[A], int]', R: 'list[list[float]]',
                        discount: float, stateFrom: S, action: A, stateTo: S, Q: 'list[list[float]]') -> float:
    transitionReward: float = R[stateIdx(stateFrom)][actionIdx(action)]
    nextStateValue: float = getStateValue(stateIdx, stateTo, Q)
    return transitionReward + discount * nextStateValue

    